### PR TITLE
Feature/address

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: app/models/address.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Address < ApplicationRecord
+  belongs_to :addressable, polymorphic: true
+
+  validates :pincode, presence: true
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/address.rb:4]: Done

[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/address.rb:4]: Done

File: app/views/users/_form.html.erb [hunk 2]
```diff
@@ -12,6 +12,9 @@
     <%= form.label :email, class:"form-label"%>
     <%= form.email_field :email, class:"form-control" %>
   </div>
+  <div class="container mt-4">
+    <%= render partial: 'shared/address', locals: { form: form, address: user.address ?  user.address : user.build_address } %>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/users/_form.html.erb:16]: Done

[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/users/_form.html.erb:16]: Changed to this

File: app/views/registrations/_form.html.erb [hunk 3]
```diff
@@ -12,6 +12,9 @@
     <%= form.label :email, class:"form-label"%>
     <%= form.email_field :email, class:"form-control" %>
   </div>
+  <div class="container mt-4">
+    <%= render partial: 'shared/address', locals: { form: form, address: user.address ?  user.address : user.build_address } %>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/registrations/_form.html.erb:16]: Done

File: app/views/branches/_form.html.erb [hunk 4]
```diff
@@ -12,6 +12,7 @@
     <%= form.label :closing_time, class: "form-label" %>
     <%= form.time_field :closing_time, class: "form-control" %>
   </div>
+  <%= render partial: 'shared/address', locals: { form: form, address: branch.address ?  branch.address : branch.build_address } %>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/branches/_form.html.erb:15]: Done

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -48,6 +48,6 @@ def send_password_reset_email
   end
 
   private def user_params
-    params.require(:user).permit(:first_name, :last_name, :role, :email, :password, :password_confirmation)
+    params.require(:user).permit(:first_name, :last_name, :role, :email, :password, :password_confirmation, address_attributes: [:address_line_1,:address_line_2, :city, :state, :pincode, :_destroy])
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/users_controller.rb:51]: Done

File: app/controllers/branches_controller.rb [hunk 6]
```diff
@@ -40,6 +40,6 @@ def destroy
   end
 
   private def branch_params
-    params.require(:branch).permit(:name, :default, :opening_time, :closing_time)
+    params.require(:branch).permit(:name, :default, :opening_time, :closing_time, address_attributes: [:address_line_1,:address_line_2, :city, :state, :pincode, :_destroy])
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/branches_controller.rb:43]: Done

File: db/migrate/20230517072848_create_addresses.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+class CreateAddresses < ActiveRecord::Migration[7.0]
+  def change
+    create_table :addresses do |t|
+      t.string :address_line_1
+      t.string :address_line_2
+      t.string :city
+      t.string :state
+      t.string :pincode
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply db/migrate/20230517072848_create_addresses.rb:8]: Done

File: app/models/address.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/address.rb:4]: Done

File: app/controllers/branches_controller.rb [hunk 8]
```diff
@@ -40,6 +40,6 @@ def destroy
   end
 
   private def branch_params
-    params.require(:branch).permit(:name, :default, :opening_time, :closing_time)
+    params.require(:branch).permit(:name, :default, :opening_time, :closing_time, address_attributes: [:address_line_1, :address_line_2, :city, :state, :pincode, :_destroy])
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/branches_controller.rb:43]: We are not allowing destroying branch address.
Fix this

File: app/models/branch.rb [hunk 9]
```diff
@@ -1,14 +1,35 @@
 # Model class Branch
 class Branch < ApplicationRecord
+  ### ASSOCIATIONS ###
+
+  has_one :address, as: :addressable, dependent: :destroy
+  accepts_nested_attributes_for :address, update_only: true, allow_destroy: true
```
Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:6]: can we remove `allow_destroy ` here as well ?

File: app/models/branch.rb [hunk 10]
```diff
@@ -1,14 +1,35 @@
...
   end
+
+  private def validate_single_default_branch
+    if default_changed? && default_branch_exists?
+      errors.add(:default, :present)
+      throw(:abort)
+    end
+  end
+
+  private def default_branch_exists?
```
Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:32]: Shouldn't this be a class method?

Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:32]: You can create a class method to return default branch.

File: app/models/branch.rb [hunk 11]
```diff
@@ -1,14 +1,35 @@
...
   private def only_one_default
     if default_was != true && Branch.exists?(default: true)
       errors.add(:default, :present)
     end
   end
+
+  private def validate_single_default_branch
+    if default_changed? && default_branch_exists?
+      errors.add(:default, :present)
+      throw(:abort)
```
Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:28]: Why `throw(:abort) ` here ?

File: app/models/branch.rb [hunk 12]
```diff
@@ -1,14 +1,35 @@
...
+
+  before_update :validate_single_default_branch
+
   private def only_one_default
     if default_was != true && Branch.exists?(default: true)
       errors.add(:default, :present)
     end
   end
+
+  private def validate_single_default_branch
```
Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:25]: Why have you created this callback?
we already have a validation for this.

File: app/views/shared/_address.html.erb [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  </div>
+  <div class="mb-3">
+    <%= address_form.label :state, class:"form-label"%>
+    <%= address_form.text_field :state, class:"form-control" %>
+  </div>
+  <div class="mb-3">
+    <%= address_form.label :pincode, class:"form-label"%>
+    <%= address_form.number_field :pincode, class:"form-control" %>
+  </div>
+  <% if address.id && address.addressable_type == 'User' %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/shared/_address.html.erb:22]: Create a method on user and address or wherever you are going to use address.
Check there if the address is destroyable or not.
Then call that method on `addressable`

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: app/controllers/branches_controller.rb (hunk 8 above)
Comment (rishabh-vinsol) [reply app/controllers/branches_controller.rb:43]: Done

File: app/models/branch.rb (hunk 10 above)
Comment (rishabh-vinsol) [reply app/models/branch.rb:32]: Done

File: app/models/branch.rb (hunk 12 above)
Comment (rishabh-vinsol) [reply app/models/branch.rb:25]: Yes fixed it

File: app/views/shared/_address.html.erb (hunk 13 above)
Comment (rishabh-vinsol) [reply app/views/shared/_address.html.erb:22]: Done

File: app/models/branch.rb (hunk 11 above)
Comment (rishabh-vinsol) [reply app/models/branch.rb:28]: Fixed

**Pull request merged**